kernel/src/syscall/mod.rs

```diff
@@ -274,7 +274,16 @@ macro_rules! impl_syscall_nums_and_dispatch_fn {
                 $(
                     $num => {
                         $crate::log_syscall_entry!($name);
-                        $crate::syscall::dispatch_fn_inner!(args, ctx, user_ctx, $handler $args)
+                        use ostd::panic;
+                        use crate::Errno;
+                        use crate::error::Error;
+                        let result = panic::catch_unwind(|| {
+                            $crate::syscall::dispatch_fn_inner!(args, ctx, user_ctx, $handler $args)
+                        });
+                        match result {
+                            Ok(syscall_return) => syscall_return,
+                            Err(_) => Err(Error::with_message(Errno::ENOSYS, "not implemented")),
```

Comment:
**Correctness concern.** After this change, even panicking triggered by `unwrap` (not `todo!` or `unimplemented!`) will be caught as `ENOSYS`.

Reply:
It will not catch `unwrap` or `panic` or `debug_assert`. 
Only the panics marked as `should_handled` in `oops.rs` will be caught 
(which is identified by the panic msg, the `todo!` is actually `panic("not yet implemented")` and `unimplemented!` is `panic("not implemented")`), 
the other panics will trigger the `panic::abort();` directly.

Reply:
The conditions is OR'ed.

```rust
        let should_handle = (!panic_on_oops && info.can_unwind())
            || format!("{}", message).contains("not yet implemented")
            || format!("{}", message).contains("not implemented");
```

Reply:
If `(!panic_on_oops && info.can_unwind())` is true, then it
> Raise the panic and expect it to be caught.
Catch it at the system call entry should be OK?

@junyang-zh Any suggestions?

Reply:
Here is the Linux's behaviors on oops.

> When the kernel detects a problem, it kills any offending processes and prints an oops message, which Linux kernel engineers can use in debugging the condition that created the oops and fixing the underlying programming error.
> From Wikipedia

This is also our intended behaviors. So when a kernel thread panics for reasons like a failed `unwrap`, we would like to kill the kernel thread. And if the thread happens to belong to a user process, kill the process, too. This way, we control the damage of a Rust panic within the thread or process, not the whole kernel.

Calling `catch_unwind` at the system call entry point catches not only panics due to `todo!` and `unimplemented!`, but also those due to a kernel logic error (like a failed `unwrap`). This is not what we want.